### PR TITLE
Correctly handle `assert_*` intrinsics

### LINKED ISSUE #6: Implement intrinsic `assert_uninit_valid`

### LINKED ISSUE #7: Implement intrinsic `assert_zero_valid`

### PR DESCRIPTION
**Pull request by adpaco-aws**
### Description of changes: 

Adds support for the `assert_uninit` and `assert_zero` intrinsics. Inserts the code for them and `assert_inhabited` into a single function.

### Resolved issues:

Resolves #6
Resolves #7 

### Call-outs:

 * All backends check the condition for `assert_inhabited` first, then proceed to do the other ones. This proposal follows the same pattern.
 * The actual error for `assert_inhabited` (i.e., in `src/test/rmc/Intrinsics/Assert/inhabited_panic.rs`) is
```
[main.assertion.1] line 13 a panicking function std::mem::MaybeUninit::<T>::assume_init is invoked: FAILURE
```
This is due to the `Nevers` hook being applied instead of the `Intrinsics` one. I have tried doing some changes to the application condition (for example, return `false` if the instance is an intrinsic) to no success (ended up in some code that expected the `Nevers` hook to have applied).
 * Some reorganization in tests. The top level folder `Assert` now contains the multiple assertions test only, and another `Assert` folder has been created within `Intrinsics` to keep the tests related to the intrinsics in this PR.

### Testing:

* How is this change tested? Existing regression + 3 new tests + 1 removed test.

* Is this a refactor change? No.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by adpaco-aws** [comment]
@zhassan-aws brought #648 to my attention, thanks!!

Removing/Changing `output_of_instance_is_never` seems the way to go, but it appears that our backend relies on it in a few places so I do not think it will be trivial. I could create a tracking issue for the `inhabited.rs` test and come back to check once the change has taken place.

File: src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs [hunk 1]
```diff
@@ -567,6 +548,48 @@ impl<'tcx> GotocCtx<'tcx> {
         )
     }
 
+    /// Generates either a panic or no-op for `assert_*` intrinsics.
+    /// These are intrinsics that statically compile to panics if the type
+    /// layout is invalid so we get a message that mentions the offending type.
```
Comment (danielsn) [inline src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs:553]: Reference to rust documentation here?

File: src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs [hunk 2]
```diff
@@ -567,6 +548,48 @@ impl<'tcx> GotocCtx<'tcx> {
...
+        // Then we check if the type allows "raw" initialization for the cases
+        // where memory is zero-initialized or entirely uninitialized
+        if intrinsic == "assert_zero_valid" && !layout.might_permit_raw_init(self, true) {
+            return self.codegen_fatal_error(
+                &format!("attempted to zero-initialize type `{}`, which is invalid", ty),
+                span,
+            );
+        }
+
+        if intrinsic == "assert_uninit_valid" && !layout.might_permit_raw_init(self, false) {
```
Comment (danielsn) [inline src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs:581]: Is there a spec we can point to here?

File: src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs [hunk 3]
```diff
@@ -567,6 +548,48 @@ impl<'tcx> GotocCtx<'tcx> {
...
+        if layout.abi.is_uninhabited() {
+            return self.codegen_fatal_error(
+                &format!("attempted to instantiate uninhabited type `{}`", ty),
+                span,
+            );
+        }
+
+        // Then we check if the type allows "raw" initialization for the cases
+        // where memory is zero-initialized or entirely uninitialized
+        if intrinsic == "assert_zero_valid" && !layout.might_permit_raw_init(self, true) {
```
Comment (danielsn) [inline src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs:574]: Is there a spec we can point to here?

File: src/test/rmc/Intrinsics/Assert/inhabited_panic.rs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// rmc-verify-fail
+
+#![feature(never_type)]
+use std::mem::MaybeUninit;
+
+// The code below attempts to instantiate uninhabited type `!`.
+// This should cause the intrinsic `assert_inhabited` to generate a panic during
+// compilation, but at present it triggers the `Nevers` hook instead.
```
Comment (danielsn) [inline src/test/rmc/Intrinsics/Assert/inhabited_panic.rs:10]: Do we have a tracking issue for this?

File: src/test/rmc/Intrinsics/Assert/zero_valid_panic.rs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// rmc-verify-fail
+
+#![feature(core_intrinsics)]
+use std::intrinsics;
+
+// The code below attempts to zero-initialize type `&i32`, causing the intrinsic
+// `assert_zero_valid` to generate a panic during compilation.
+fn main() {
```
Comment (danielsn) [inline src/test/rmc/Intrinsics/Assert/zero_valid_panic.rs:10]: Should we also have a test with an expected pass (e.g. zero initializing a `u8`)?

**Review by danielsn — COMMENTED** [review]
The tests I see are for primitive types. Would it make sense to also ensure that we do the right thing for enums and structs?

File: src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs (hunk 1 above)
Comment (adpaco-aws) [reply src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs:553]: Added links to rust docs.

File: src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs (hunk 2 above)
Comment (adpaco-aws) [reply src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs:584]: I was not able to find one as documentation regarding these intrinsics is quite hard to find.

The closest thing I could find is the actual implementation in `rustc_codegen_ssa/src/mir/block.rs` (`codegen_panic_intrinsic`) in addition to PR discussions. But if you know of one I will be happy to add it here.

File: src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs (hunk 3 above)
Comment (adpaco-aws) [reply src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs:577]: See above.

File: src/test/rmc/Intrinsics/Assert/inhabited_panic.rs (hunk 4 above)
Comment (adpaco-aws) [reply src/test/rmc/Intrinsics/Assert/inhabited_panic.rs:10]: Created #751 which I linked here.

File: src/test/rmc/Intrinsics/Assert/zero_valid_panic.rs (hunk 5 above)
Comment (adpaco-aws) [reply src/test/rmc/Intrinsics/Assert/zero_valid_panic.rs:10]: The test in `src/test/rmc/Intrinsics/Assert/zero_valid.rs` (path updated in this PR) already does this for several types, including a user-defined struct.

File: src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs (hunk 2 above)
Comment (tedinski) [reply src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs:584]: FWIW, I think links directly to code serve as a reasonable reference, too. Maybe add that as a comment? (e.g. link directly to it on github?)

**Review by tedinski — APPROVED** [review]

File: src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs (hunk 2 above)
Comment (adpaco-aws) [reply src/rmc-compiler/rustc_codegen_rmc/src/codegen/intrinsic.rs]: My main concern with that is that links to `rustc` code may become outdated pretty quickly. But even that is better that no links, so just added a note here linking to that function.

**Pull request merged**